internal/pkg/agent/application/upgrade/manual_rollback.go

```diff
@@ -0,0 +1,258 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package upgrade
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/filelock"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/paths"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/reexec"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/upgrade/details"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+	"github.com/elastic/elastic-agent/internal/pkg/release"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+	"github.com/elastic/elastic-agent/pkg/version"
+	agtversion "github.com/elastic/elastic-agent/version"
+)
+
+const disableRollbackWindow = time.Duration(0)
+
+func (u *Upgrader) rollbackToPreviousVersion(ctx context.Context, topDir string, now time.Time, version string, action *fleetapi.ActionUpgrade) (reexec.ShutdownCallbackFn, error) {
+	if version == "" {
+		return nil, ErrEmptyRollbackVersion
+	}
+
+	// check that the upgrade marker exists and is accessible
+	updateMarkerPath := markerFilePath(paths.DataFrom(topDir))
+	_, err := os.Stat(updateMarkerPath)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return nil, fmt.Errorf("stat() on upgrade marker %q failed: %w", updateMarkerPath, err)
+	}
+
+	var watcherExecutable string
+	var versionedHomeToRollbackTo string
+	var updateMarkerExistsBeforeRollback bool
+
+	if errors.Is(err, os.ErrNotExist) {
+		// there is no upgrade marker, we need to extract available rollbacks from agent installs
+		watcherExecutable, versionedHomeToRollbackTo, err = rollbackUsingAgentInstalls(u.log, u.watcherHelper, u.availableRollbacksSource, topDir, now, version, u.markUpgrade)
+	} else {
+		updateMarkerExistsBeforeRollback = true
+		watcherExecutable, versionedHomeToRollbackTo, err = rollbackUsingUpgradeMarker(ctx, u.log, u.watcherHelper, topDir, now, version)
```

Comment:
Just wondering if we need this case.  Couldn't we always `rollbackUsingAgentInstalls`?

Reply:
Fixed in 52ab103953